File: Design-Patterns-Using-JavaScript/Decorator/DecoratorPatternWindow.js
// Window (Component)
class Window {
    draw() {
        throw new Error("Abstract method: draw");
    }
}

// SimpleWindow (ConcreteComponent)
class SimpleWindow extends Window {
    draw() {
        console.log("SimpleWindow draw.");
    }
}

// Decorator (Decorator)
class Decorator extends Window {
    constructor(component) {
        super();
        this.component = component;
    }

    draw() {
        this.component.draw();
    }
}

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator extends Decorator {
    constructor(component) {
        super(component);
    }

    draw() {
        super.draw();
        console.log("VerticalScrollBarDecorator draw");
    }
}

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator extends Decorator {
    constructor(component) {
        super(component);
    }

    draw() {
        super.draw();
        console.log("HorizontalScrollBarDecorator draw");
    }
}

// Client code
const component = new SimpleWindow();
const decorator1 = new VerticalScrollBarDecorator(component);
const decorator2 = new HorizontalScrollBarDecorator(decorator1);
decorator2.draw();

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw
*/

File: Design-Patterns-Using-Swift/Decorator/DecoratorPatternWindow.swift
// Window (Component)
protocol Window {
    func draw()
}

// SimpleWindow (ConcreteComponent)
class SimpleWindow: Window {
    func draw() {
        print("SimpleWindow draw.")
    }
}

// Decorator (Decorator)
class Decorator: Window {
    var component: Window

    init(_ component: Window) {
        self.component = component
    }

    func draw() {
        component.draw()
    }
}

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator: Decorator {
    override func draw() {
        super.draw()
        print("VerticalScrollBarDecorator draw")
    }
}

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator: Decorator {
    override func draw() {
        super.draw()
        print("HorizontalScrollBarDecorator draw")
    }
}

// Client code
let component: Window = SimpleWindow()
let decorator1: Window = VerticalScrollBarDecorator(component)
let decorator2: Window = HorizontalScrollBarDecorator(decorator1)
decorator2.draw()

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw
*/
